src/ripple/app/main/GRPCServer.cpp

```diff
@@ -0,0 +1,410 @@
+            "gRPC-Client",
+            [thisShared](std::shared_ptr<JobQueue::Coro> coro) {
+                std::lock_guard<std::mutex> lock(thisShared->mut_);
+
+                // Do nothing if call has been aborted due to server shutdown
+                if (thisShared->aborted_)
+                    return;
+
+                thisShared->process(coro);
+                thisShared->finished_ = true;
+            });
+    }
+    else
+    {
+        BOOST_ASSERT(false);
+    }
+}
+
+template <class Request, class Response>
+void
+GRPCServerImpl::CallData<Request, Response>::process(
+    std::shared_ptr<JobQueue::Coro> coro)
+{
+    try
+    {
+        auto usage = getUsage();
+        if (usage.disconnect())
+        {
+            grpc::Status status{grpc::StatusCode::RESOURCE_EXHAUSTED,
+                                "usage balance exceeds threshhold"};
+            responder_.FinishWithError(status, this);
+        }
+        else
+        {
+            auto loadType = getLoadType();
+            usage.charge(loadType);
+            auto role = getRole();
+
+            RPC::GRPCContext<Request> context{{app_.journal("gRPCServer"),
+                                               app_,
+                                               loadType,
+                                               app_.getOPs(),
+                                               app_.getLedgerMaster(),
+                                               usage,
+                                               role,
+                                               coro,
+                                               InfoSub::pointer()},
+                                              request_};
+
+            // Make sure we can currently handle the rpc
+            error_code_i conditionMetRes =
+                RPC::conditionMet(requiredCondition_, context);
+
+            if (conditionMetRes != rpcSUCCESS)
+            {
+                RPC::ErrorInfo errorInfo = RPC::get_error_info(conditionMetRes);
+                grpc::Status status{grpc::StatusCode::INTERNAL,
+                                    errorInfo.message.c_str()};
+                responder_.FinishWithError(status, this);
+            }
+            else
+            {
+                std::pair<Response, grpc::Status> result = handler_(context);
+                responder_.Finish(result.first, result.second, this);
+            }
+        }
+    }
+    catch (std::exception const& ex)
+    {
+        grpc::Status status{grpc::StatusCode::INTERNAL, ex.what()};
+        responder_.FinishWithError(status, this);
+    }
+}
+
+template <class Request, class Response>
+bool
+GRPCServerImpl::CallData<Request, Response>::isFinished()
+{
+    // checking the status while a request is in the middle of being
+    // processed will lead to indeterminate results. Lock here to
+    // sequence checking status and processing
+    std::lock_guard<std::mutex> lock(mut_);
+    return finished_;
```

Comment:
We actually need the lock because this `CallData` object can be returned from `cq_.Next()` as soon as the response is sent, and `finished_` is set to true after the response is sent. Without the lock, there is a race condition, where the object could be returned from `cq_.Next()` before `finished_` is actually set to true. I updated the comment to explain this situation better.

L84 is really just a debugging sanity check, to make sure the handler is not executed twice. As the code is, it's not needed because we call `isFinished()` inside `handleRpcs()` before calling `process()`. However, the assert has stopped me from making certain errors during development, so I'd like to keep it. I don't really want to lock just for a sanity check though. I changed the code a bit to make it more clear that this is just a sanity check.

I also added a check for `finished_` inside the coroutine, to cover the case where the coroutine is executed twice **_concurrently_** ( as opposed to executed once to completion, and then executed again). I overlooked this case before.